www/docs/server/data-transformers.md

```diff
@@ -51,11 +50,73 @@ export default trpcNext.createNextApiHandler({
 });
 ```
 
+## Different transformers for upload and download
+
+If a transformer should only be used for one directon or different transformers should be used for upload and download (e.g. for performance reasons), you can provide individual transformers for upload and download. Make sure you use the same combined transformer everywhere.
+
+### How to
+
+Here [superjson](https://github.com/blitz-js/superjson) is be used for uploading and [devalue](https://github.com/Rich-Harris/devalue) for downloading data, because devalue is a lot faster but insecure to use on the server.
+
+#### 0. Install
+
+```bash
+yarn add superjson devalue
+```
+
+#### 1. Add to `utils/trpc.ts`
+
+```ts
+import superjson from 'superjson';
+import devalue from 'devalue';
+
+// [...]
+
+export const transformer = {
+  input: superjson,
+  output: {
+    serialize: (object) => devalue(object),
+    deserialize: (object) => eval(`(${object})`),
+  },
+};
+```
```

Comment:
The only thing I don't like with this is that `devalue` isn't actually needed at all on the client and becomes an unnecessary import - got any idea to work around this?

Reply:
I guess on the client you only need `input.serialize` & `output.deserialize` and the inverse for the server.... 🤔

Reply:
At least when using devalue this is no problem. Devalue has no dependencies and should be very small.

Reply:
```tsx
export type CombinedDataTransformerClient = {
  input: Pick<DataTransformer, 'serialize'>;
  output: Pick<DataTransformer, 'deserialize'>;
};
```

Reply:
The types (on the trpcclient) should still reflect that you can omit it if you want so it's up to the dev to optimize to the level they want

Reply:
I've added `CombinedDataTransformerClient` type and used in client packages, but I am still not sure if this will prevent devalue to be imported on client.

Reply:
> The types (on the trpcclient) should still reflect that you can omit it if you want so it's up to the dev to optimize to the level they want

Ah, I see. The type is added in https://github.com/trpc/trpc/pull/393/commits/9845d012a7e7bef1f1f77b591a72f61ea85b20ea